### ISSUE TITLE
Possibility to save the gherkin keywords (Given/When/Then/And etc..) to the meta information 

### ISSUE THREAD

**Issue by rian-777** [Arthy000/gherkin-testcafe#99]
Hi @Arthy000 ,

I was wondering if it would be possible to save the gherkin keywords as well to the test meta in addition to type & text ?

I see that the keyword information is already available in the gherkinDocument and if it could be saved to the meta, it would help with custom reporter implementations without the additional keyword handle code.

**Comment by Arthy000** [comment]
Hello @rian-777 

I wanted to do that at first, using the keyword type wasn't my first choice.
However when you look at the pickle object provided by Cucumber for each scenario:
```js
{
  id: '84ca28b3-d813-46ae-b738-c87650189347',
  uri: '/Users/arthurwarnier/gherkin-testcafe/examples/google.feature',
  astNodeIds: [
    '6771cdc0-cc64-459b-b337-51ac24d49bc5',
    '53875f0c-c727-4a6b-888e-f05220208c2b'
  ],
  name: 'Searching for twitter on Google',
  language: 'en',
  steps: [
    {
      id: 'a1974f4d-3064-4d53-8bfb-a022cd3571e1',
      text: "I open Google's search page",
      type: 'Context',
      argument: undefined,
      astNodeIds: [Array]
    },
    {
      id: '4173d779-f666-46ec-a347-07d4011e7868',
      text: 'I dismiss the privacy statement when it appears',
      type: 'Action',
      argument: undefined,
      astNodeIds: [Array]
    },
    {
      id: '045221de-81e3-4051-8878-facee1f2d4b3',
      text: 'I am typing my search request "twitter" on Google',
      type: 'Action',
      argument: undefined,
      astNodeIds: [Array]
    },
    {
      id: '2834df52-69a3-44e8-97fb-8d03af937921',
      text: 'I am pressing "enter" key on Google',
      type: 'Action',
      argument: undefined,
      astNodeIds: [Array]
    },
    {
      id: '31b4faaa-31b4-4aa8-9e38-163e4a8c1d4c',
      text: `I should see that the first Google's result is "Twitter"`,
      type: 'Outcome',
      argument: undefined,
      astNodeIds: [Array]
    }
  ],
  tags: []
} 
```
you can see that the keywork itself is in fact not present.

**Comment by rian-777** [comment]
Hi @Arthy000, thank you for the response as well as for this amazing project.

At the time of opening this issue, my initial thought process was to replace the existing forEach,  to loop through the gherkinDocument.feature.children for every scenario so that it would have the keyword information as well. 

But, on more tests, I found that my approach caused a problem when the scenario has a background (not getting executed as background itself is a different object which may/may not be present and would need additional code handling) or when the gherkin step pattern has some parameter (text would be incorrect)

So from an execution standpoint, your current implementation of looping through the pickle is indeed ideal :)

From the test meta perspective though, the 'And' keyword always gets translated into type "Action" in pickle but the children.scenario.step in gherkinDocument would show it correctly as "Conjunction". This was also the reason why I thought of getting details from gherkinDocument in the first place.

Do you think it would still be possible to extract just keyword information from children.background.step / children.scenario.step as I see that the step.id in the gherkinDocument matches the step.astNodeIds in the pickle?

**Comment by Arthy000** [comment]
@rian-777 I'll look I to it, thanks for the investigation :)

**Comment by Arthy000** [comment]
@rian-777 I'll look I to it, thanks for the investigation :)

**Comment by rian-777** [comment]
Hi @Arthy000 , Just checking in,  if you were able to take a look into this suggestion.

I had done some tests locally using the comparison logic mentioned in my previous comment - step.astNodeIds[0] in the pickle vs step.id in children.background.step / children.scenario.step and it is indeed possible to get information as needed, not sure of the code performance aspect though as I am pretty new to javascript.

Also, since my actual use case was to create a customized report , I also explored some of the existing testcafe-reporter packages available. I found that for easy integration with existing reporter packages, it would be good to have a "**name**" attribute for each step in the meta, which will contain the entire keyword+text information (for background, i was thinking to prefix the text with Background: ") 

For example:

 steps: [
    {
      astNodeIds: [Array],
      text: "I open Google's search page",
      name: "Background: Given I open Google's search page"
    },
    {
      astNodeIds: [Array],
      text: 'I am typing my search request "github TestCafe" on Google',
      name: 'Given I am typing my search request "github TestCafe" on Google'
    },
]

Please let me know your thoughts on this idea of customizing metadata information

**Comment by Arthy000** [comment]
Hi @rian-777, I had a very busy week, I'm going to work on this in the next few days, I'll keep you posted.

**Comment by Arthy000** [comment]
Hi, I've made progress, I'm just trying to figure out an efficient way to handle background steps now, once that's done I'll addapt the existing reporters to showcase how to use it

**Comment by Arthy000** [comment]
Hello @rian-777 
I've just created a [new branch](https://github.com/Arthy000/gherkin-testcafe/tree/feature/add-step-keywords-metadata), could you check if it matches your expectations?

**Comment by Rinu-EYGit** [comment]
Hi @Arthy000 , the recent changes that you have added is exactly the information that I wanted, in the test meta :)

Regarding the name property, I am in agreement with what you have mentioned. Since we have the prefix, keyword and text information are now in the meta, we can make use of these as needed while implementing custom reporter packages.

Thanks for considering this suggestion and the quick turn around. Really appreciate it 👍

**Comment by rian-777** [comment]
Hi @Arthy000 , the latest changes are exactly the information that I was looking for in test meta :)

I also agree regarding your comments on the name prop, which was basically a  'nice to have'. Since, we have the prefix, keyword and text information already in place, the customization within reporters can be taken care as per requirement.

Thanks for implementing this suggestion and the quick turnaround on the same. Really appreciate it 👍

**Comment by Arthy000** [comment]
Published in version 5.5.0
